### DISCUSSION TITLE
Help about langchain with LLM

### DISCUSSION THREAD

**Discussion by woo320** [community/community#82737] [Programming Help]
### Body

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain import HuggingFacePipeline
from langchain import HuggingFacePipeline, PromptTemplate
from langchain.chains import LLMChain, RetrievalQA
from langchain.vectorstores import Pinecone
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
import pinecone
import os

loader = PyPDFLoader('재난법2.pdf')
data = loader.load()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
texts = text_splitter.split_documents(data)

embeddings = HuggingFaceEmbeddings(
    model_name="jhgan/ko-sroberta-multitask", encode_kwargs={'normalize_embeddings': True}
)

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, eos_token_id=tokenizer.eos_token_id)
llm = HuggingFacePipeline(
    pipeline = pipe,
    model_kwargs={"temperature": 0.2, "max_length": 512},
)

PINECONE_API_KEY = os.environ.get("PINECONE_API_KEY", "981c772f-74d3-42d1-9158-e301f20af75e")
PINECONE_API_ENV = os.environ.get("PINECONE_API_ENV", "gcp-starter")

pinecone.init(api_key=PINECONE_API_KEY, environment=PINECONE_API_ENV)
index_name = "langchainpinecone"

docsearch=Pinecone.from_texts([t.page_content for t in texts], embeddings, index_name=index_name)

query = "자연재난이 뭐야?"
docs = docsearch.search(query, search_type='similarity', k=2)

SYSTEM_PROMPT = """Use following pieces of context to answer the question in Korean.
"""

B_INST, E_INST = "[INST]", "[/INST]"
B_SYS, E_SYS = "<<<SYS>>>\n", "\n<<</SYS>>>\n\n"

SYSTEM_PROMPT = B_SYS + SYSTEM_PROMPT + E_SYS

instruction = """
Convert the following text from English to Korean:\n {context}

Convert the following text from English to Korean:\n {question}
"""

template = B_INST + SYSTEM_PROMPT + instruction + E_INST

prompt = PromptTemplate(template=template, input_variables=["context", "question"])

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=docsearch.as_retriever(search_kwargs={"k": 2}),
    return_source_documents=True,
    chain_type_kwargs={"prompt": prompt},
)

result = qa_chain("긴급구조기관이 뭐야?")
result['result']

My goal is to learn the pdf file and get an answer in Korean.
Please help me.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by rishiskhare** [comment]
Consider modifying the instruction variable to use Korean itself:
instruction = """
다음 텍스트를 영어에서 한국어로 번역하세요:\n {context}

다음 텍스트를 영어에서 한국어로 번역하세요:\n {question}
"""

template = B_INST + SYSTEM_PROMPT + instruction + E_INST

prompt = PromptTemplate(template=template, input_variables=["context", "question"])

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=docsearch.as_retriever(search_kwargs={"k": 2}),
    return_source_documents=True,
    chain_type_kwargs={"prompt": prompt},
)

result = qa_chain("긴급구조기관이 뭐야?")
result['result']

In this way, the model would be more likely to output Korean because its instructions are given in Korean.

Please mark this answer as correct if this was helpful :)

**Reply by woo320** [reply]
The answers are still coming out in English. Is there any other way? Please help..

**Reply by rishiskhare** [reply]
Sure! I found this: https://huggingface.co/quantumaikr/llama-2-70b-fb16-korean

Consider using this fine-tuned Llama model for Korean.

Use the following script:
from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain import HuggingFacePipeline
from langchain import HuggingFacePipeline, PromptTemplate
from langchain.chains import LLMChain, RetrievalQA
from langchain.vectorstores import Pinecone
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
import pinecone
import os

loader = PyPDFLoader('재난법2.pdf')
data = loader.load()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
texts = text_splitter.split_documents(data)

embeddings = HuggingFaceEmbeddings(
model_name="jhgan/ko-sroberta-multitask", encode_kwargs={'normalize_embeddings': True}
)

tokenizer = AutoTokenizer.from_pretrained("quantumaikr/llama-2-70b-fb16-korean")
model = AutoModelForCausalLM.from_pretrained("quantumaikr/llama-2-70b-fb16-korean", torch_dtype=torch.float16, device_map="auto")
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, eos_token_id=tokenizer.eos_token_id)
llm = HuggingFacePipeline(
pipeline = pipe,
model_kwargs={"temperature": 0.2, "max_length": 512},
)

PINECONE_API_KEY = os.environ.get("PINECONE_API_KEY", "981c772f-74d3-42d1-9158-e301f20af75e")
PINECONE_API_ENV = os.environ.get("PINECONE_API_ENV", "gcp-starter")

pinecone.init(api_key=PINECONE_API_KEY, environment=PINECONE_API_ENV)
index_name = "langchainpinecone"

docsearch=Pinecone.from_texts([t.page_content for t in texts], embeddings, index_name=index_name)

query = "자연재난이 뭐야?"
docs = docsearch.search(query, search_type='similarity', k=2)

SYSTEM_PROMPT = """Use following pieces of context to answer the question in Korean.
"""

B_INST, E_INST = "[INST]", "[/INST]"
B_SYS, E_SYS = "<<>>\n", "\n<<>>\n\n"

SYSTEM_PROMPT = B_SYS + SYSTEM_PROMPT + E_SYS

instruction = """
다음 텍스트를 영어에서 한국어로 번역하세요:\n {context}

다음 텍스트를 영어에서 한국어로 번역하세요:\n {question}
"""

template = B_INST + SYSTEM_PROMPT + instruction + E_INST

prompt = PromptTemplate(template=template, input_variables=["context", "question"])

qa_chain = RetrievalQA.from_chain_type(
llm=llm,
chain_type="stuff",
retriever=docsearch.as_retriever(search_kwargs={"k": 2}),
return_source_documents=True,
chain_type_kwargs={"prompt": prompt},
)

result = qa_chain("긴급구조기관이 뭐야?")
result['result']

**Comment by MamoNew** [comment]
The answers are still coming out in English. Is there any other way?
